### PR TITLE
Rescoring with 4-gram LM

### PR DESCRIPTION
**Pull request by csukuangfj**
~Support for rescoring with the whole lattice is not added yet, as k2 does not support removing epsilon self-loops at present.~

Here are the WERs (%) using LM rescoring. You can see that rescoring with LM reduces WER.

## Base line (MMI conformer training)
It uses `k2.shortest_path` for decoding
- test_clean: 6.80
- test_other: 18.03

## With LM rescoring
It uses `k2.random_paths` with various `num_paths`

Screen Shot 2021-04-05 at 1 23 36 PM

### CODE REVIEW

File: snowfall/decoding/lm_rescore.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    # num_repeats.num_elements() == unique_word_seqs.num_elements()
+    #
+    # Since k2.ragged.unique_sequences will reorder paths within a seq,
+    # `new2old` is a 1-D torch.Tensor mapping from the output path index
+    # to the input path index.
+    # new2old.numel() == unique_word_seqs.num_elements()
+    unique_word_seqs, num_repeats, new2old = k2.ragged.unique_sequences(
+        word_seqs, need_num_repeats=True, need_new2old_indexes=True)
+
+    seq_to_path_shape = k2.ragged.get_layer(unique_word_seqs.shape(), 0)
```
Comment (pzelasko) [inline snowfall/decoding/lm_rescore.py:127]: side note: I think it would be very helpful to have a manual that describes these methods and how FSAs/Ragged/etc. can be manipulated, in a similar way to what numpy has (https://numpy.org/doc/stable/user/basics.indexing.html)

File: snowfall/decoding/lm_rescore.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                                  path_to_seq_map)
+
+    # Now compute lm_scores
+    b_to_a_map = torch.zeros_like(path_to_seq_map)
+    lm_path_lats = k2.intersect_device(G,
+                                       word_fsas_with_epsilon_loops,
+                                       b_to_a_map=b_to_a_map,
+                                       sorted_match_a=True)
+    lm_path_lats = k2.top_sort(k2.connect(lm_path_lats.to('cpu'))).to(device)
+    lm_scores = lm_path_lats.get_tot_scores(True, True)
```
Comment (pzelasko) [inline snowfall/decoding/lm_rescore.py:153]: what is the meaning of `(True, True)`? could these be keyword args instead?

File: snowfall/decoding/lm_rescore.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+    tot_scores = am_scores + lm_scores
```
Comment (pzelasko) [inline snowfall/decoding/lm_rescore.py:155]: doesn't this way of rescoring remove the "graph scores" that come from e.g. multiple entries with distinct probabilities in the lexicon?

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py [hunk 4]
```diff
@@ -288,6 +312,40 @@ def main():
...
         HLG = k2.Fsa.from_dict(d)
 
+    if args.use_lm_rescoring:
+        logging.debug(f'Rescoring with num_paths: {num_paths}')
+        first_word_disambig_id = find_first_disambig_symbol(symbol_table)
+        logging.debug('Loading G_4_gram')
+        if not os.path.exists(lang_dir / 'G_4_gram.pt'):
+            with open(lang_dir / 'G_4_gram.fst.txt') as f:
+                G = k2.Fsa.from_openfst(f.read(), acceptor=False)
+                del G.aux_labels
```
[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py:322]: You can view an existing 3-gram G at https://github.com/csukuangfj/kaldilm/blob/master/G.svg
or by visiting https://github.com/csukuangfj/kaldilm

The resulting G is a transducer, not an acceptor. What we want is an acceptor, so we remove its `aux_labels` here.
If we don't remove it, its `aux_labels` will be propagated in later operations.

File: snowfall/decoding/lm_rescore.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+
+    # The `scores` of every arc consists of `am_scores` and `lm_scores`
+    am_path_lats.scores = am_path_lats.scores - am_path_lats.lm_scores
+
+    am_scores = am_path_lats.get_tot_scores(True, True)
+
+    return am_scores
+
+
+@torch.no_grad()
```
[the remark this answers is not in the document]
Comment (csukuangfj) [reply snowfall/decoding/lm_rescore.py:63]: We are not using gradient in decoding.

`no_grad` can be removed in this case. It will not affect the result. But I believe it requires more memory as it
needs to save the graph information for backward.

I am not sure how gradient will be used in the decoding script. That's why I am using `no_grad` here.

File: snowfall/decoding/lm_rescore.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    assert hasattr(lats, 'aux_labels')
+    assert hasattr(lats, 'lm_scores')
+
+    assert G.shape == (1, None, None)
+    assert G.device == device
+    assert hasattr(G, 'aux_labels') is False
+
+    # First, extract `num_paths` paths for each sequence.
+    # paths is a k2.RaggedInt with axes [seq][path][arc_pos]
+    paths = k2.random_paths(lats, num_paths=num_paths, use_double_scores=True)
```
[the remark this answers is not in the document]
Comment (csukuangfj) [reply snowfall/decoding/lm_rescore.py:100]: `k2.random_paths` is a template function in C++ and it has instantiations for both `float` and `double`.

`use_doube_scores` means it will use `double`, instead of `float` in C++.
The precision of `double` is higher than `float`, so I am using `double` here.

You can use `use_double_scores=False`, but I am not sure how it will affect the result.

File: snowfall/decoding/lm_rescore.py (hunk 1 above)
Comment (csukuangfj) [reply snowfall/decoding/lm_rescore.py:127]: I agree. 

We definitely need to improve the documentation from the Python side.

I guess the current documention in Python makes the most sense only to those who are familiar with k2 C++ APIs.

File: snowfall/decoding/lm_rescore.py (hunk 2 above)
Comment (csukuangfj) [reply snowfall/decoding/lm_rescore.py:153]: The documentation about these two arguments can be found at
https://k2.readthedocs.io/en/latest/python_api/fsa.html#id2


`get_tot_scores` accepts two args:
-  use_double_scores:  This argument has the same meaning as its counterpart in `k2.random_paths`
- log_semiring: If it is true, it uses log semiring; otherwise, it uses tropical semiring.

> could these be keyword args instead?

Yes, I should have used keyword args here. But since both of them are True, so I just used positional arguments.

File: snowfall/decoding/lm_rescore.py (hunk 3 above)
Comment (csukuangfj) [reply snowfall/decoding/lm_rescore.py:155]: I believe we are currently using a lexicon with log-prob 0 on all arcs, so the lexicon does not contribute probabilities to the final rescoring.

The current approach assumes that LM scores come exclusively from the language model.

**Comment by pzelasko** [comment]
@csukuangfj thanks for clarifying!

File: snowfall/decoding/lm_rescore.py (hunk 3 above)
Comment (csukuangfj) [reply snowfall/decoding/lm_rescore.py:155]: What I am currently doing is

- Get an n-best list.  (Note that the `best` in `n-best` is not the same one as in Kaldi. This list is sampled from a lattice)
- For each list (i.e., for each path), we remove the scores of very arc that come from the language model (i.e., HLG).
- Get the acoustic scores of each path
- Get the LM scores of each path
- Get the total scores of each path, with is am_scores + lm_scores
- The path with the largest total scores is the best path, which is used as the decoding output

Comment (csukuangfj) [reply snowfall/decoding/lm_rescore.py:155]: What I am currently doing is

- Get an n-best list.  (Note that the `best` in `n-best` is not the same one as in Kaldi. This list is sampled from a lattice)
- For each list (i.e., for each path), we remove the scores of very arc that come from the language model (i.e., HLG).
- Get the acoustic scores of each path
- Get the LM scores of each path
- Get the total scores of each path, with is am_scores + lm_scores
- The path with the largest total scores is the best path, which is used as the decoding output

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py [hunk 7]
```diff
@@ -288,6 +312,40 @@ def main():
...
+                G = k2.arc_sort(G)
+                torch.save(G.as_dict(), lang_dir / 'G_4_gram.pt')
+        else:
+            logging.debug('Loading pre-compiled G_4_gram')
+            d = torch.load(lang_dir / 'G_4_gram.pt')
+            G = k2.Fsa.from_dict(d).to(device)
+
+        if num_paths is None:
+            # TODO(fangjun): remove this
+            assert False, 'Rescoring with whole lattice is currently not supported'
```
[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py:337]: Will implement it once https://github.com/k2-fsa/k2/pull/705  is ready.

**Comment by csukuangfj** [comment]
Here is the result for rescoring with the whole lattice.
(NOTE: It needs https://github.com/k2-fsa/k2/pull/705 to test this pull-request)

Screen Shot 2021-04-06 at 5 46 39 PM

You can see that rescoring with the whole lattice has a lower WER than that of n-best list, even if it has a smaller output beam.


---

I have to decrease the [output beam][1] size in `k2.intersect_dense_pruned` from  7 to `2` and `1` to avoid the following error.

```
[F] /ceph-fj/fangjun/open-source/s2/k2/k2/csrc/intersect.cu:1054:void k2::DeviceIntersector::ForwardSortedA() Problem size is too larg
e for this code: b_state_bits=18, key_bits=46, value_bits=18, value_max=262144, next_state_end=268477

```



[1]: https://github.com/k2-fsa/snowfall/blob/1255c3a287939701226d412ae1150add719dac3a/egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py#L75

**Comment by danpovey** [comment]
Great!!
To-morrow I'll finish fixing the hash thing so that limit won't be a problem.

**Comment by csukuangfj** [comment]
Here are updated WERs with a larger output beam using changes from https://github.com/k2-fsa/k2/pull/708

The WER for test-clean is decreased from 6.80% to 5.81%.

The WER is likely to further decrease with a larger output beam. We will see after Dan changes the implementation of hash in k2.

Screen Shot 2021-04-07 at 10 59 59 AM

**Comment by danpovey** [comment]
Great!!  Will do it now...

File: snowfall/decoding/lm_rescore.py [hunk 8]
```diff
@@ -217,10 +217,30 @@ def rescore_with_whole_lattice(lats: k2.Fsa,
...
+                                             sorted_match_a=True)
+    except RuntimeError as e:
+        print(f'Caught exception:\n{e}\n')
+        print(f'Number of FSAs: {inverted_lats.shape[0]}')
+        print('num_arcs before pruning: ',
+              inverted_lats_with_epsilon_loops.arcs.num_elements())
+
+        # NOTE(fangjun): The choice of the threshold 0.01 is arbitrary here
+        # to avoid OOM. We may need to fine tune it.
+        inverted_lats = k2.prune_on_arc_post(inverted_lats, 0.001, True)
```
Comment (csukuangfj) [inline snowfall/decoding/lm_rescore.py:233]: It needs https://github.com/k2-fsa/k2/pull/713

**Comment by csukuangfj** [comment]
Here are the latest results.

Ready to review and merge.

### Average over last 5 epochs

#### LM rescoring with whole lattice

```
$ ./mmi_att_transformer_decode.py --use-lm-rescoring=1 --num-path=-1 --max-duration=10 --output-beam-size=8

2021-04-11 10:37:58,913 INFO [common.py:356] [test-clean] %WER 5.72% [3008 / 52576, 562 ins, 164 del, 2282 sub ]
2021-04-11 10:46:03,670 INFO [common.py:356] [test-other] %WER 15.71% [8224 / 52343, 1331 ins, 562 del, 6331 sub ]
```

#### LM rescoring with n-best list

```
$ ./mmi_att_transformer_decode.py --use-lm-rescoring=1 --num-path=100 --max-duration=500 --output-beam-size=20

2021-04-11 15:17:07,792 INFO [common.py:356] [test-clean] %WER 6.31% [3316 / 52576, 746 ins, 160 del, 2410 sub ]
2021-04-11 15:19:48,583 INFO [common.py:356] [test-other] %WER 16.93% [8863 / 52343, 1649 ins, 514 del, 6700 sub ]
```

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py]: Fixed by removal.

File: egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py [hunk 9]
```diff
@@ -206,6 +224,12 @@ def main():
     max_duration = args.max_duration
     avg = args.avg
     att_rate = args.att_rate
+    use_lm_rescoring = args.use_lm_rescoring
+    num_paths = args.num_paths
+    if num_paths < 0:
+        num_paths = None
+    else:
+        assert num_paths > 1
```
[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/librispeech/asr/simple_v1/mmi_att_transformer_decode.py]: A new bool variable `use_whole_lattice` is introduced and `num_paths` is not changed anymore.

**Comment by danpovey** [comment]
Great!!
When we do RNNLM rescoring with n-best lists, we'll have to figure out an
algorithm that is more accurate.
Probably it will involve a couple passes of rescoring paths.
Dan

File: snowfall/decoding/lm_rescore.py (hunk 3 above)
Comment (pzelasko) [reply snowfall/decoding/lm_rescore.py:155]: Sorry I missed this reply. My question was this: in Kaldi, HCLG arc scores actually consisted of two types of scores: the LM scores, and "graph" scores, which came from pronunciation probabilities in lexiconp. AFAIK rescoring in Kaldi always first re-computed the LM scores with the original G and subtracted them from the LM scores in the lattice, so that the "graph" scores were preserved, and only then added the rescoring LM scores on top of that.

I am not sure if our HLG supports multiple pronunciations and their probabilities (or do we even need to do it anymore), but it's just something that came to my mind when reviewing.

**Comment by danpovey** [comment]
Mm.
Currently we are not using pron-probs, although we could in the future.
We can easily attach additional cost types and store them as
differently-named attributes, if we need to in the future.
The FST-reading code already supports reading arbitrary numbers of fields.

File: snowfall/decoding/lm_rescore.py [hunk 10]
```diff
@@ -2,10 +2,52 @@
...
+
+        fsas = k2.index(b_fsas, indexes)
+        b_to_a = k2.index(b_to_a_map, indexes)
+        path_lats = k2.intersect_device(a_fsas,
+                                        fsas,
+                                        b_to_a_map=b_to_a,
+                                        sorted_match_a=sorted_match_a)
+        ans.append(path_lats)
+
+    return k2.append(ans)
```
Comment (pzelasko) [inline snowfall/decoding/lm_rescore.py:48]: just wondering - shouldn't that be named `cat/concat` to be more consistent with numpy/torch/pandas?

Comment (csukuangfj) [reply snowfall/decoding/lm_rescore.py:48]: Yes, I agree with you. I was also once confused by this name. Only after I read its implementation did I realize it's actually equivalent to cat.

The current name is to match its counterpart in C++. Don't know what's Dan's opinion about this. @danpovey

**Review by danpovey — APPROVED** [review]
I think this should be good to merge!

**Comment by danpovey** [comment]
I am fine to change both of them to cat.
